### PR TITLE
Implement FLO & TXD Instructions on GPU Shaders

### PR DESCRIPTION
**Pull request by FernandoS27**
This implements two new instructions: FLO and TXD.

FLO: calculates the Most Significant Bit of an input
TXD: loads from a texture using derivates.

### CODE REVIEW

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 1]
```diff
@@ -1772,6 +1810,14 @@ class GLSLDecompiler final {
         return {tmp, Type::Float};
     }
 
+    Expression TextureGradient(Operation operation) {
+        const auto meta = std::get_if<MetaTexture>(&operation.GetMeta());
+        ASSERT(meta);
+
+        std::string expr = GenerateTexture(operation, "Grad", {TextureDerivates{}, TextureAoffi{}});
+        return {expr + GetSwizzle(meta->element), Type::Float};
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]:
```suggestion
        return {std::move(expr) + GetSwizzle(meta->element), Type::Float};
```

File: src/video_core/shader/decode/arithmetic_integer.cpp [hunk 2]
```diff
@@ -130,6 +130,24 @@ u32 ShaderIR::DecodeArithmeticInteger(NodeBlock& bb, u32 pc) {
...
+        }
+        if (instr.flo.is_signed) {
+            value = Operation(OperationCode::IBitMSB, NO_PRECISE, op_b);
+        } else {
+            value = Operation(OperationCode::UBitMSB, NO_PRECISE, op_b);
+        }
+        if (instr.flo.sh) {
+            value = Operation(OperationCode::UBitwiseXor, NO_PRECISE, value, Immediate(31));
+        }
+        SetRegister(bb, instr.gpr0, value);
```
Comment (lioncash) [inline src/video_core/shader/decode/arithmetic_integer.cpp]:
```suggestion
        SetRegister(bb, instr.gpr0, std::move(value));
```

File: src/video_core/shader/decode/arithmetic_integer.cpp [hunk 3]
```diff
@@ -130,6 +130,24 @@ u32 ShaderIR::DecodeArithmeticInteger(NodeBlock& bb, u32 pc) {
...
+        if (instr.flo.invert) {
+            op_b = Operation(OperationCode::IBitwiseNot, NO_PRECISE, op_b);
+        }
+        if (instr.flo.is_signed) {
+            value = Operation(OperationCode::IBitMSB, NO_PRECISE, op_b);
+        } else {
+            value = Operation(OperationCode::UBitMSB, NO_PRECISE, op_b);
+        }
+        if (instr.flo.sh) {
+            value = Operation(OperationCode::UBitwiseXor, NO_PRECISE, value, Immediate(31));
```
Comment (lioncash) [inline src/video_core/shader/decode/arithmetic_integer.cpp]:
```suggestion
            value = Operation(OperationCode::UBitwiseXor, NO_PRECISE, std::move(value), Immediate(31));
```

File: src/video_core/shader/decode/arithmetic_integer.cpp [hunk 4]
```diff
@@ -130,6 +130,24 @@ u32 ShaderIR::DecodeArithmeticInteger(NodeBlock& bb, u32 pc) {
...
     }
+    case OpCode::Id::FLO_R:
+    case OpCode::Id::FLO_C:
+    case OpCode::Id::FLO_IMM: {
+        Node value;
+        if (instr.flo.invert) {
+            op_b = Operation(OperationCode::IBitwiseNot, NO_PRECISE, op_b);
+        }
+        if (instr.flo.is_signed) {
+            value = Operation(OperationCode::IBitMSB, NO_PRECISE, op_b);
```
Comment (lioncash) [inline src/video_core/shader/decode/arithmetic_integer.cpp]:
```suggestion
            value = Operation(OperationCode::IBitMSB, NO_PRECISE, std::move(op_b));
```

File: src/video_core/shader/decode/arithmetic_integer.cpp [hunk 5]
```diff
@@ -130,6 +130,24 @@ u32 ShaderIR::DecodeArithmeticInteger(NodeBlock& bb, u32 pc) {
...
+        } else {
+            value = Operation(OperationCode::UBitMSB, NO_PRECISE, op_b);
```
Comment (lioncash) [inline src/video_core/shader/decode/arithmetic_integer.cpp]:
```suggestion
            value = Operation(OperationCode::UBitMSB, NO_PRECISE, std::move(op_b));
```

File: src/video_core/shader/decode/arithmetic_integer.cpp [hunk 6]
```diff
@@ -130,6 +130,24 @@ u32 ShaderIR::DecodeArithmeticInteger(NodeBlock& bb, u32 pc) {
         SetRegister(bb, instr.gpr0, value);
         break;
     }
+    case OpCode::Id::FLO_R:
+    case OpCode::Id::FLO_C:
+    case OpCode::Id::FLO_IMM: {
+        Node value;
+        if (instr.flo.invert) {
+            op_b = Operation(OperationCode::IBitwiseNot, NO_PRECISE, op_b);
```
Comment (lioncash) [inline src/video_core/shader/decode/arithmetic_integer.cpp]:
```suggestion
            op_b = Operation(OperationCode::IBitwiseNot, NO_PRECISE, std::move(op_b));
```

File: src/video_core/shader/decode/texture.cpp [hunk 7]
```diff
@@ -134,13 +134,55 @@ u32 ShaderIR::DecodeTexture(NodeBlock& bb, u32 pc) {
...
+            derivates.push_back(GetRegister(derivate_reg + derivate));
+            derivates.push_back(GetRegister(derivate_reg + derivate + 1));
+        }
+
+        Node4 values;
+        for (u32 element = 0; element < values.size(); ++element) {
+            auto coords_copy = coords;
+            MetaTexture meta{sampler, {}, {}, {}, derivates, {}, {}, {}, element};
+            values[element] =
+                Operation(OperationCode::TextureGradient, meta, std::move(coords_copy));
```
Comment (lioncash) [inline src/video_core/shader/decode/texture.cpp]:
```suggestion
                Operation(OperationCode::TextureGradient, std::move(meta), std::move(coords_copy));
```

File: src/video_core/shader/decode/texture.cpp [hunk 8]
```diff
@@ -134,13 +134,55 @@ u32 ShaderIR::DecodeTexture(NodeBlock& bb, u32 pc) {
...
+        for (std::size_t i = 0; i < coord_count; ++i) {
+            coords.push_back(GetRegister(base_reg + i));
+            const std::size_t derivate = i * 2;
+            derivates.push_back(GetRegister(derivate_reg + derivate));
+            derivates.push_back(GetRegister(derivate_reg + derivate + 1));
+        }
+
+        Node4 values;
+        for (u32 element = 0; element < values.size(); ++element) {
+            auto coords_copy = coords;
```
Comment (lioncash) [inline src/video_core/shader/decode/texture.cpp]: Is there a reason you don't just remove the std::move in the `Operation` call and pass `coords` by value instead? This seems to be unnecessary.

File: src/video_core/shader/decode/texture.cpp [hunk 9]
```diff
@@ -134,13 +134,55 @@ u32 ShaderIR::DecodeTexture(NodeBlock& bb, u32 pc) {
...
         WriteTexsInstructionFloat(bb, instr, values, true);
         break;
     }
+    case OpCode::Id::TXD_B:
+        is_bindless = true;
+        [[fallthrough]];
+    case OpCode::Id::TXD: {
+        UNIMPLEMENTED_IF_MSG(instr.txd.UsesMiscMode(TextureMiscMode::AOFFI),
+                             "AOFFI is not implemented");
+        const auto is_array = static_cast<bool>(instr.txd.is_array != 0);
```
Comment (ReinUsesLisp) [inline src/video_core/shader/decode/texture.cpp]: This is unused and is contradictory with the assert below it.

File: src/video_core/shader/decode/texture.cpp (hunk 8 above)
Comment (FernandoS27) [reply src/video_core/shader/decode/texture.cpp]: Most of the code is just following the rest of the code in that file which does that. I also had my doubts.

**Pull request merged**